rust: fix splicing for rust hooks

DavHau opened the pull request:
This fixes a long standing issue where rust hooks behave differently when used inside buildRustPackage vs inside mkDerivation, which lead to surprising behavior, like for example the package being built for the wrong paltform or the linker not being found especially in cross compilation scenarios.

The reason for this inconsitency was, that buildRustPackage consumed the hooks in a non-spliced form, via [this inherit statement](https://github.com/NixOS/nixpkgs/blob/4506ece030a0c82d078edd0360ea8af6b4d94035/pkgs/development/compilers/rust/make-rust-platform.nix#L60), and therefore the usual platform shift on the hooks introduced by putting them in `nativeBuildInputs` was not applied here.

Thoug whenever the hook was used inside other builders like `mkDerivation` the platform shift did apply correctly as the hook was consumed via the spliced package set, introducing the inconsitecy.

Because of the non-spliced use in buildRustPackage, most rust hooks have been designed with the wrong build/host/target shift in mind which is fixed by this change.

Due to the inconsitent behavior between different builders, workarounds like `rust.envVars`, which were previously introduced, likely become obsolete by this change.

This likely fixes a bunch of cross compilation issues for rust packages that are not based on `buildRustPackage` but instead consume the hooks directly.

Done:
- ensure that `buildRustPackage` consumes spliced hooks by using makeScopeWithSplicing' in make-rust-platform.nix.
- refactor hooks to make them refer to correct build/host/target packages.
- remove `rust.envVars` workaround from all rust hooks
- implement tests for most rust hooks in /pkgs/test/rust-hooks

The newly added tests can be executed for native as well as cross system via:
```
nix-build -A tests.rust-hooks -A pkgsCross.riscv64.tests.rust-hooks
```


<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

alyssais commented:
The diff is very big and hard to review. Can you please break this up into multiple commits, especially the reformatting?

DavHau committed:
format pkgs/build-support/rust/hooks/default.nix

DavHau committed:
rust: fix splicing for rust hooks

This fixes a long standing issue where rust hooks behave differently when used inside buildRustPackage vs inside mkDerivation, which lead to surprising behavior, like for example the package being built for the wrong paltform or the linker not being found especially in cross compilation scenarios.

The reason for this inconsitency was, that buildRustPackage consumed the hooks in a non-spliced form, via [this inherit statement](https://github.com/NixOS/nixpkgs/blob/4506ece030a0c82d078edd0360ea8af6b4d94035/pkgs/development/compilers/rust/make-rust-platform.nix#L60), and therefore the usual platform shift on the hooks introduced by putting them in `nativeBuildInputs` was not applied here.

Thoug whenever the hook was used inside other builders like `mkDerivation` the platform shift did apply correctly as the hook was consumed via the spliced package set, introducing the inconsitecy.

Because of the wrong (non-spliced) use in buildRustPackage, most rust hooks have been designed with the wrong build/host/target shift in mind which is fixed by this change.

Due to the inconsitent behavior between different builders, workarounds like `rust.envVars`, which were previously introduced, likely become obsolete by this change.

This likely fixes a bunch of cross compilation issues for rust packages that are not based on `buildRustPackage` but instead consume the hooks directly.

Done:
- ensure that `buildRustPackage` consumes spliced hooks by using makeScopeWithSplicing' in make-rust-platform.nix.
- refactor hooks to make them refer to correct build/host/target packages.
- remove `rust.envVars` workaround from all rust hooks
- implement tests for most rust hooks in /pkgs/test/rut-hooks

The newly added tests can be executed for native as well as cross compilation via:
```
nix-build -A tests.rust-hooks -A pkgsCross.riscv64.tests.rust-hooks
```

DavHau commented:
The formatting is now a separate commit

DavHau commented on pkgs/development/compilers/rust/make-rust-platform.nix:
```diff
@@ -17,59 +18,64 @@
-  buildRustPackage = callPackage ../../../build-support/rust/build-rust-package {
-    inherit
-      stdenv
-      cargoBuildHook
-      cargoCheckHook
-      cargoInstallHook
-      cargoNextestHook
-      cargoSetupHook
-      fetchCargoTarball
-      fetchCargoVendor
-      importCargoLock
```
This was the root cause of the inconsistency. Since these hooks were consumed directly in their non-spliced form here, they did not get their platform shifts applied inside buildRustPackage, which made the hooks behave differently here than in other packages where they were consumed via `rustPlatform.xxx`.

simonzkl commented on pkgs/development/compilers/rust/make-rust-platform.nix:
```diff
@@ -17,59 +18,64 @@
-    git = buildPackages.gitMinimal;
-    inherit cargo;
-  };
+(makeScopeWithSplicing' {
+  otherSplices = generateSplicesForMkScope "rustPlatform";
```
This locks down `makeRustPlatform` so it can only be used as `pkgs.rustPlatform`.

If you define your own custom platform with `makeRustPlatform`, then `rustc` in other splices will point to `pkgs.rustPlatform` which is incorrect.

simonzkl replied:
I think the correct thing to do here would be either:
a) Pass down `otherSplices` to the caller of `makeRustPlatform`
b) Build all `otherSplices` manually (`selfBuildBuild = lib.makeScope pkgsBuildBuild.newScope f`, ...)

danielbarter commented:
@DavHau: I think this broke cross building some packages: https://github.com/NixOS/nixpkgs/issues/384662.

`nix build .#pkgsCross.aarch64-multiplatform.helix` breaks on `db4cf6cb1dd04630efa35706050884d18a700374` but works on `06d633e659b6bf004b309018d1edfa2a7a7010ae`

Would you be able to take a look? I am not an expert, but it seems like the nix build env is not successfully communicating to `cargo` that this should be a cross build.

szlend commented:
@danielbarter Could you try if https://github.com/NixOS/nixpkgs/pull/384493 fixes the issue for you?

szlend commented:
The removal of `setEnv` unset a bunch of useful variables typically used by crates like `cc-rs`. For example `CC_<target>`/`CXX_<target>`/`HOST_CC`/`HOST_CXX`.

danielbarter commented:
@szlend: This makes sense.

I had some time to explore a bit also. Can confirm that the issue is that `cc-rs` is getting confused about what target to build for. `tree-sitter` is both a runtime dependency and a build time dependency, so it needs to be built twice.

danielbarter commented:
Can confirm, that the removal of `setEnv` is the issue. Either needs to be added back in, or we override `buildPhase` for helix with `rust.envVars.setEnv` like a few other packages are doing.

@szlend: you want to open PR with whatever you think is best?

danielbarter commented:
https://github.com/NixOS/nixpkgs/pull/386191: re adding in setEnv
